LIVY-231: Multi node HA for batch sessions

meisam opened the pull request:
This is a preliminary PR for LIVY-231 (https://issues.cloudera.org/browse/LIVY-231) and it has known issues, but we can use it to discuss the design of multi-node HA for Livy.

The PR uses a cache for each Livy node. The cache keeps sessions' metadata in sync with ZooKeeper. Any change in ZooKKeeper data updates signals cache and updates the local copy of the data on Livy nodes.

The cache is implemented using Apache curator's "Path Cache" recipe: http://curator.apache.org/curator-recipes/path-cache.html.

This PR should be revised based on https://github.com/cloudera/livy/pull/220 (JIRA ticket: https://issues.cloudera.org/browse/LIVY-239)

tc0312 commented:
We are trying to get a stable build for 0.3 first for Spark 2.0 and session recovery. Do you mind if we handle HA after 0.3 is released?

meisam commented:
That would work.

meisam committed:
LIVY-231 Session recovery for batch sessions with multiple Livy servers

1. stores the session ID in Zookeeper/Filesystem

2. builds a cache on top of Zookeeper. The cache keeps metadata of all
batch sessions across all Livy servers connected to it and notifies all
livy servers with changes in cache.

3. Adds callback methods from SessionStore to SessionManager.
SessionStore watches events in the ZooKeeper cache and calls into proper
callback methods in SessionManager.

Task-url: https://issues.cloudera.org/browse/LIVY-231

meisam committed:
LIVY-231 Updating tests for Session recovery multiple Livy servers

Mocking batchSessionsCache in ZooKeeperStateStore.
Without mocking, Livy test fails to start a ZooKeeperStateStore.

Task-url: https://issues.cloudera.org/browse/LIVY-231

meisam commented:
I updated the pull request to fix the merge errors.

shenh062326 commented:
Hi, @meisam, @alex-the-man, is there any progress of this issue. Livy ha is import to us.

prabhu1984 commented:
Hi @shenh062326,

We have enabled MultiNode HA long time back on Paypal and have been using for close to a year and we already submitted 1.3 million spark jobs through livy multinode HA. 

Even today we presented or updates on [Spark Summit](https://spark-summit.org/2017/events/spark-compute-as-a-service-at-paypal/) and committed to open source all our enhancements. @meisam  will send update PR soon and we will merge it soon.

Thanks
Prabhu

harsh829 commented:
Is there any progress on multi-node HA for interactive sessions?
